File: main/src/main/java/lavaplayer/source/soundcloud/SoundCloudSegmentDecoder.java
package lavaplayer.source.soundcloud;

import lavaplayer.tools.io.SeekableInputStream;
import lavaplayer.track.playback.AudioProcessingContext;

import java.io.IOException;
import java.util.function.Supplier;

public interface SoundCloudSegmentDecoder extends AutoCloseable {
    void prepareStream(boolean beginning) throws IOException;

    void resetStream() throws IOException;

    void playStream(
        AudioProcessingContext context,
        long startPosition,
        long desiredPosition
    ) throws InterruptedException, IOException;

    interface Factory {
        SoundCloudSegmentDecoder create(Supplier<SeekableInputStream> nextStreamProvider);
    }
}

File: main/src/main/java/lavaplayer/source/soundcloud/SoundCloudSegmentDecoder.kt
package lavaplayer.source.soundcloud

import lavaplayer.tools.io.SeekableInputStream
import lavaplayer.track.playback.AudioProcessingContext
import java.io.IOException
import java.util.function.Supplier

interface SoundCloudSegmentDecoder : AutoCloseable {
    @Throws(IOException::class)
    fun prepareStream(beginning: Boolean)

    @Throws(IOException::class)
    fun resetStream()

    @Throws(InterruptedException::class, IOException::class)
    fun playStream(context: AudioProcessingContext, startPosition: Long, desiredPosition: Long)

    fun interface Factory {
        fun create(nextStreamProvider: Supplier<SeekableInputStream>): SoundCloudSegmentDecoder
    }
}
